Add maximum mean discrepancy and radial basis

kevinchern opened the pull request:
TODOs:

- [x] document
- [x] test
- [x] release note
- [x] wrap with `store_config`

kevinchern reviewed (commented):
Thanks Vlad!! Nicely implemented and documented.
Let's add the MaximumMeanDiscrepancy a a module and this should be good to go.

thisac commented on dwave/plugins/torch/models/losses/mmd.py:
```diff
@@ -0,0 +1,183 @@
+        Returns:
+            torch.Tensor: A (n, n) tensor.
+        """
+        pass
```
No need for pass if there is a docstring.
```suggestion
```

thisac commented on dwave/plugins/torch/models/losses/mmd.py:
```diff
@@ -0,0 +1,183 @@
+        Returns:
+            torch.Tensor: A (n_x + n_y, n_x + n_y) tensor.
+        """
+        assert x.shape[1:] == y.shape[1:], (
```
Should be an if-check with an exception raised instead of an assert.

VolodyaCO replied:
Replaced it

thisac commented on tests/test_dvae_winci2020.py:
```diff
@@ -84,6 +86,28 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            soft = logits
+            result = hard - soft.detach() + soft
+            # Now we need to repeat the result n_samples times along a new dimension
+            return repeat(result, "b ... -> b n ...", n=n_samples)
```
Do we absolutely need `repeat` here? Seems a bit cumbersome to add einops as a test dependency just for this test. 🤔

thisac reviewed (commented):
Thanks @VolodyaCO and @kevinchern! Test fail due to Python 3.9 tests being run (removing 3.9 support in #49, which should fix it).

Unit tests should be expanded. There should at least be test classes for kernels (`RBF`) and more unit tests for the `mmd` function. Otherwise, looks good. Just a few, mostly minor, comments.

VolodyaCO commented on dwave/plugins/torch/models/losses/mmd.py:
```diff
@@ -0,0 +1,183 @@
+        return self._kernel(xy)
+
+
+class RBFKernel(Kernel):
```
Done!

VolodyaCO commented on dwave/plugins/torch/models/losses/mmd.py:
```diff
@@ -0,0 +1,183 @@
+
+class Kernel(nn.Module):
+    """
+    Base class for kernels.
```
Done !

VolodyaCO commented on dwave/plugins/torch/models/losses/mmd.py:
```diff
@@ -0,0 +1,183 @@
+        Args:
+            x (torch.Tensor): A (n_x, f1, f2, ...) tensor.
+            y (torch.Tensor): A (n_y, f1, f2, ...) tensor.
+        Returns:
```
Done

kevinchern commented:
Just did another pass. Main changes are:

1. Refactor MMD into kernels, functional, and loss modules.
2. Rename acronyms to be consistent with package standards (I'm torn between adhering to our no-acronym standard versus conventions used in pytorch, i.e., pytorch uses `MSELoss`, but we define `MaximuMeanDiscrepancyLoss`)
3. Fixed bug with RBF (need to flatten before `torch.cdist`)
4. Wrote unit tests
5. `_kernel(x)` changed to `_kernel(x, y)`.
6. Added specific errors: `class SampleSizeError(ValueError): ...` and `class DimensionMismatchError(ValueError): ...`

thisac commented on dwave/plugins/torch/nn/modules/kernels.py:
```diff
@@ -0,0 +1,145 @@
+    (n, f1, f2, ...), where n is the number of items and f1, f2, ... are feature dimensions, so that
+    the output is a tensor of shape (n, n) containing the pairwise kernel values.
+    """
+    @abstractmethod
```
Cannot use abstract methods in non-abstract classes (or rather, this won't do anything).

thisac commented on dwave/plugins/torch/nn/modules/loss.py:
```diff
@@ -0,0 +1,56 @@
+
+
+class MaximumMeanDiscrepancyLoss(nn.Module):
+    """An unbiased estimator for the squared maximum mean discrepancy as a loss function.
```
Suggestion, since MMD is referred to in the forward method below.
```suggestion
    """An unbiased estimator for the squared Maximum Mean Discrepancy (MMD) as a loss function.
```

thisac commented on dwave/plugins/torch/nn/functional.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def maximum_mean_discrepancy_loss(x: torch.Tensor, y: torch.Tensor, kernel: Kernel) -> torch.Tensor:
+    """Estimates the squared maximum mean discrepancy (MMD) given two samples ``x`` and ``y``.
```
Shouldn't capitalized acronyms have capitalized expansions? 🤔 
```suggestion
    """Estimates the squared Maximum Mean Discrepancy (MMD) given two samples ``x`` and ``y``.
```

thisac reviewed (commented):
Overall, looks good! Just a few comments/suggestions.

thisac commented on dwave/plugins/torch/nn/modules/loss.py:
```diff
@@ -0,0 +1,56 @@
+    """
+
+    @store_config
+    def __init__(self, kernel: Kernel):
```
```suggestion
    def __init__(self, kernel: Kernel) -> None:
```
